src/ejabberd_cowboy.erl

```diff
@@ -224,3 +227,68 @@ maybe_insert_max_connections(TransportOpts, Opts) ->
             NewTuple = {Key, Value},
             lists:keystore(Key, 1, TransportOpts, NewTuple)
     end.
+
+-spec measured_methods() -> [mongoose_cowboy_metrics:method()].
+measured_methods() ->
+    [<<"GET">>,
+     <<"HEAD">>,
+     <<"POST">>,
+     <<"PUT">>,
+     <<"DELETE">>,
+     <<"OPTIONS">>,
+     <<"PATCH">>].
+
+-spec measured_classes() -> [mongoose_cowboy_metrics:status_class()].
+measured_classes() ->
+    [<<"1XX">>, <<"2XX">>, <<"3XX">>, <<"4XX">>, <<"5XX">>].
+
+base_metrics_prefix() ->
+    [http].
+
+middlewares_with_metrics() ->
+    [mongoose_cowboy_metrics_mw_before,
+     cowboy_router,
+     mongoose_cowboy_metrics_mw_after,
+     cowboy_handler].
+
+-spec maybe_init_metrics(list()) -> {MetricsEnv :: list(), MetricsProtocolOpts :: list()}.
+maybe_init_metrics(Opts) ->
+    case proplists:get_value(metrics, Opts, false) of
+        true ->
+            BasePrefix = base_metrics_prefix(),
+            HandlerToPrefixMappings = build_metric_prefixes(
+                                          BasePrefix, proplists:get_value(modules, Opts, []), #{}),
+            [create_metrics(Prefix) || Prefix <- maps:values(HandlerToPrefixMappings)],
+            {[{record_metrics, true}, {handler_to_metric_prefix, HandlerToPrefixMappings}],
+             [{middlewares, middlewares_with_metrics()}]};
+        false ->
+            {[], []}
+    end.
+
+-spec build_metric_prefixes(BasePrefix :: list(), Modules :: [tuple()], Acc) -> Acc
+    when Acc :: #{module() => mongoose_cowboy_metrics:prefix()}.
+build_metric_prefixes(_BasePrefix, [], Acc) ->
+    Acc;
+build_metric_prefixes(BasePrefix, [{_Host, _Path, Handler, _HandlerOpts, Opts} | Tail], Acc) ->
```

Comment:
Yes, I agree that in general case it won't work well. However, in our HTTP API handlers usually do the exact same thing given some HTTP method (i.e. it doesn't matter what was the path or parameters). 

The other solution would be hand-crafted metrics, different for each handler (i.e. we explicitly report metrics from each HTTP module). However, we would not get any metrics in case the handler has crashed. 

Maybe the metrics for each handler in a single `ejabberd_cowboy's` scope should opt-in as well? This way we can turn them on only for endpoints which we know it makes sense to have these kinds of metrics for.

Reply:
OK, I understand it's good enough until someone actually needs a more fine-grained names. :)